### ISSUE TITLE
while compiling cc_library from DefaultInfo bazel ignores bazelrc

### ISSUE THREAD

**Issue by onelxj** [bazelbuild/bazel#15623]
### Description of the bug:

I have a rule that returns DefaultInfo(depset(...)) - folder that consists of c++ sources and headers, then I'm trying to compile it with 

```
cc_library(
    name = "route-guide-eventuals-library",
    srcs = [
        ":route-guide-eventuals-generated",  # Default info with srcs
    ],
    hdrs = [
        ":route-guide-eventuals-generated",  # Default info with srcs
    ],
    # Can't get .bazelrc settings for generated files.
    # copts = ["-std=c++17"],
    alwayslink = True,
    deps =[
        ":route-guide-grpc",
    ]
)
```

The sources require c++17, in bazelrc I have 
```
build:linux --cxxopt='-std=c++17'
build:linux --host_cxxopt=-std=c++17
```
but whenever I try to run - it uses C++11 instead of 17
if uncomment `copts`, that includes '-std=c++17' - it works

### Which operating system are you running Bazel on?

Linux Ubuntu 22.04

### What is the output of `bazel info release`?

release 5.1.1

**Comment by onelxj** [comment]
Fixed via adding `build --action_env=BAZEL_CXXOPTS="-std=c++17"` to bazelrc

Thanks =)

**Comment by fmeum** [comment]
> Fixed via adding `build --action_env=BAZEL_CXXOPTS="-std=c++17"` to bazelrc
> 
> Thanks =)

`--repo_env` should work better as it doesn't pollute the environment of every action. 
Also, please keep in mind that what counts for the `deps` of a `cc_library` is what's in the `CcInfo` of your custom rule, not in `DefaultInfo`. If all you do is generate files, you may want to reference the custom rule from `srcs`, not `deps`.

**Comment by onelxj** [comment]
@fmeum Could you please explain, why defaul bazel:linux --cxxopt='-std=c++17' is not valid for this build? Other srcs in the repo that copliling also with cc_library work fine

**Comment by fmeum** [comment]
@onelxj Do you have a complete example reproducing the issue you could share? In a local example I tried `--cxxopt` did apply to generated files with e.g. `.cpp` extensions.

**Comment by Mizux** [comment]
Trying to force C++ dialect (std c++17) in subcommands on various platforms and using various bazel build arguments…

Project \  Platform| gLinux<br>Linux github| gMacOS (Intel/M1)| MacOS github (intel) | gWindows<br>Windows github
-- | -- | -- | -- | --
Repo:bazel-cpp<br>Branch: arg<br>–cxxopt=”-std=c++17” | APPEND | APPEND | APPEND | ADDED
Repo:bazel-cpp<br>Branch: env<br>–action_env=BAZEL_CXXPROTOS=”-std=c++17” | REPLACE | REPLACE | IGNORED | IGNORED
  |   |   |   |  
Repo:bazel-proto<br>protobuf v21.12<br>–cxxopt=”-std=c++17” | APPEND | APPEND | APPEND | ADDED
Repo:bazel-proto<br>protobuf v21.12<br>–action_env=BAZEL_CXXPROTOS=”-std=c++17” | REPLACE | REPLACE | IGNORED<br>C++11 used -> absl build fail | IGNORED
  |   |   |   |  
Repo:bazel-proto<br>Branch: arg<br>protobuf v22.2 + Fix #12320<br>–cxxopt=”-std=c++17” | IGNORED<br>C++0x used  absl build fail | IGNORED<br>C++0x used -> absl build fail | APPEND | IGNORED
Repo: bazel-proto<br>Branch: env<br>protobuf v22.2 + Fix #12320<br>–action_env=BAZEL_CXXPROTOS=”-std=c++17” | REPLACE | REPLACEbug #12306 | IGNORED<br>C++11 used -> absl build fail | IGNORED

Legend:
* `ADDED`: the c++17 flag is added to the subcommand
* `APPEND`: std=c++17 is appended in the subcommand (i.e. the default flag is still present)
* `REPLACE`: the default flag is replaced by the one passed as argument
* `IGNORED`: the flags argument seems completely ignored/dropped and I can’t find it in any subcommand trace log

dev note: use `--subcommands=true` or `-s` to get `# SUBCOMMAND ...` trace log
repo: https://mizux.dev/bazel-cpp
repo: https://mizux.dev/bazel-proto
